src/libgap.c

```diff
@@ -0,0 +1,219 @@
+/* LibGAP - a shared library version of the GAP kernel
+ * Copyright (C) 2013 Volker Braun <vbraun.name@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+ * USA.
+ */
+
+
+#include "libgap.h"
+#include "system.h"
+#include "gap.h"
+#include "gasman.h"
+#include "libgap_internal.h"
+#include "scanner.h"
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+
+/* Pointers to input/output buffers. libGAP users must not access these
+ * buffers
+ * directly! */
+
+#define BUFFER_STEP 16 * 1024
+
+static char * stdin_buffer = NULL;
+
+static char * stdout_buffer = NULL;
+static size_t stdout_bufsize = 0;
+static size_t stdout_pos = 0;
+
+
+/* stderr is captured in a static buffer to make it easier to pass it around
+ * in the error handler */
+
+#define STDERR_BUFSIZE 4096
+
+static char   stderr_buffer[STDERR_BUFSIZE];
+static size_t stderr_pos = 0;
+
+int libgap_in_enter_exit_block = 0; /* false */
+
+/*************************************************************************/
+/*** Initialize / Finalize ***********************************************/
+/*************************************************************************/
+
+/* from gap.c */
+extern char ** sysenviron;
+extern Int     NrImportedGVars;
+extern Int     NrImportedFuncs;
+
+
+/*************************************************************************/
+/*** Global Initialization ***********************************************/
+/*************************************************************************/
+
+void libgap_initialize(int argc, char ** argv)
+{
+    sysenviron = environ;
+
+    /* Init interpreter state */
+    fprintf(stderr, "init global state\n");
+    InitMainGlobalState();
+
+    NrImportedGVars = 0;
+    NrImportedFuncs = 0;
+
+    /* This won't work if we do an HPC-GAP version of libgap */
+    TLS(UserHasQUIT) = 0;
+    TLS(UserHasQuit) = 0;
+    libgap_mark_stack_bottom();
+    InitializeGap(&argc, argv);
+}
```

Comment:
How about we move this code into `InitializeGap` instead?
```
  sysenviron = environ;
  NrImportedGVars = 0;
  NrImportedFuncs = 0;
  TLS(UserHasQUIT) = 0;
  TLS(UserHasQuit) = 0;
```
And then of course modify GAP's `main` function accordingly, too. This way, we don't need to turn `NrImportedGVars` etc. into globally visible variabls.